### PR TITLE
Game Manager

### LINKED ISSUE #24: FEAT | Game Manager class

### PR DESCRIPTION
**Pull request by louisgv**
## Issues resolved:

- [ ] closes #24

### CODE REVIEW

File: Assets/Developers/Lab/Components/DayNight.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  {
+    DAY = 0,
+    NIGHT = 1
+  }
+
+  namespace Data {
+    [Serializable]
+    public struct DayNight : IComponentData
+    {
+      public DayNightPhase Phase;
```
Comment (DennisSSDev) [inline Assets/Developers/Lab/Components/DayNight.cs:22]: Not using E for Enum naming convention?

File: Assets/Developers/Lab/Components/Move.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using Unity.Collections;
+using Unity.Entities;
+using Unity.Mathematics;
+using UnityEngine;
+
+namespace JanusVagabond
+{
+  public enum MoveDirection
```
Comment (DennisSSDev) [inline Assets/Developers/Lab/Components/Move.cs:9]: Same as before

File: Assets/Developers/Lab/Components/Move.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  namespace Data
+  {
+
+    [Serializable]
+    public struct Move : IComponentData
+    {
+      public float Speed;
+      public float MaxSpeed;
+      public float MinSpeed;
+      // public MoveDirection Direction;
```
Comment (DennisSSDev) [inline Assets/Developers/Lab/Components/Move.cs:39]: Remove commented out code

File: Assets/Developers/Lab/Systems/DayNightSystem.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    // [BurstCompile]
+    struct SpawnJob : IJobForEachWithEntity<Data.DayNight, LocalToWorld>
+    {
+      // Add fields here that your job needs to do its work.
+      // For example,
+      //    public float deltaTime;
+
+      public EntityCommandBuffer.Concurrent CommandBuffer;
+
+      public void Execute(Entity entity, int index, [ReadOnly] ref Data.DayNight data,
```
Comment (DennisSSDev) [inline Assets/Developers/Lab/Systems/DayNightSystem.cs:30]: This system doesn't look like it's doing much. But I think we discussed this during the call, so I expect you're working on it

File: Assets/Developers/Lab/Systems/MovementSystem.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+using Random = Unity.Mathematics.Random;
+
+namespace JanusVagabond
+{
+  public class MovementSystem : ComponentSystem
+  {
+    const int RANDOM_SEED = 99;
+
+    static Random m_RandomPool = new Random(RANDOM_SEED);
+    protected override void OnUpdate()
```
Comment (DennisSSDev) [inline Assets/Developers/Lab/Systems/MovementSystem.cs:15]: It is an `override` func, but you never call base.OnUpdate(), is there anything important that parent OnUpdate calls?

File: Assets/Developers/Lab/Systems/MovementSystem.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
+﻿using Unity.Entities;
+using Unity.Mathematics;
+using Unity.Transforms;
+using UnityEngine;
+
+using Random = Unity.Mathematics.Random;
+
+namespace JanusVagabond
+{
+  public class MovementSystem : ComponentSystem
```
Comment (DennisSSDev) [inline Assets/Developers/Lab/Systems/MovementSystem.cs:10]: It's a little bit unclear what MovementSystem really does. It seems that it affects multiple different entities with different types of components. I would add a comment to make it more clear what it does :^)

File: Assets/Developers/Lab/Systems/SpawnSystem.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    static Random m_RandomPool = new Random(RANDOM_SEED);
+
+    BeginInitializationEntityCommandBufferSystem m_EntityCommandBufferSystem;
+    protected override void OnCreate()
+    {
+      // Cache the BeginInitializationEntityCommandBufferSystem in a field, so we don't have to create it every frame
+      m_EntityCommandBufferSystem = World.GetOrCreateSystem<BeginInitializationEntityCommandBufferSystem>();
+    }
+
+    // [BurstCompile]
```
Comment (DennisSSDev) [inline Assets/Developers/Lab/Systems/SpawnSystem.cs:25]: Not sure why you commented out all the `BurstCompile`s. Is there a reason for it? I would probably remove them if the feature doesn't work

**Review by DennisSSDev — APPROVED** [review]
LGTM. I have minor change requests and clarifications. It would also be great if you could answer some of the questions for tracking purposes

File: Assets/Developers/Lab/Components/DayNight.cs (hunk 1 above)
Comment (louisgv) [reply Assets/Developers/Lab/Components/DayNight.cs:22]: That's the UE4 C++ convention. For C# reference MSDN: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members

File: Assets/Developers/Lab/Components/Move.cs (hunk 3 above)
Comment (louisgv) [reply Assets/Developers/Lab/Components/Move.cs:39]: Will need this eventually but I'm not using it yet. Will add a TODO comment here

File: Assets/Developers/Lab/Systems/MovementSystem.cs (hunk 5 above)
Comment (louisgv) [reply Assets/Developers/Lab/Systems/MovementSystem.cs:15]: The OnUpdate is called by the main system itself (i.e a kind of scheduler). It's similar to the Update or Start of MonoBehavior.

File: Assets/Developers/Lab/Systems/MovementSystem.cs (hunk 6 above)
Comment (louisgv) [reply Assets/Developers/Lab/Systems/MovementSystem.cs:10]: Sounds good.

To explain why it affects multiple type of component, note that a system is freely to query any type of entity. I.e, MovementSystem can query entity from DayNightComponent. We just organize it like this at the moment for simple logic, but usually a System can query anything in the environment to implement its logic, as long as the logic is complete.

I.e, DayNight can query for monster spawner to spawn monster at night for example.

File: Assets/Developers/Lab/Systems/SpawnSystem.cs (hunk 7 above)
Comment (louisgv) [reply Assets/Developers/Lab/Systems/SpawnSystem.cs:25]: Talked a bit about this during last call. The BurstCompiler doesn't work with Job that deals with Transform for some reason (only LocalToWorld). Just leaving it there to eventually enable it once they support more type. Will add a comment on it.

**Pull request merged**